akka-stream/src/main/scala/akka/stream/scaladsl/FlowWithContext.scala

```diff
@@ -0,0 +1,119 @@
+
+  def collect[Out2](f: PartialFunction[Out, Out2]): Repr[Ctx, Out2] =
+    via(flow.collect {
+      case (e, ctx) if f.isDefinedAt(e) ⇒ (f(e), ctx)
+    })
+
+  def filter(pred: Out ⇒ Boolean): Repr[Ctx, Out] =
+    collect { case e if pred(e) ⇒ e }
+
+  def filterNot(pred: Out ⇒ Boolean): Repr[Ctx, Out] =
+    collect { case e if !pred(e) ⇒ e }
+
+  def grouped(n: Int): Repr[immutable.Seq[Ctx], immutable.Seq[Out]] =
+    via(flow.grouped(n).map { elsWithContext ⇒
+      val (els, ctxs) = elsWithContext.unzip
+      (els, ctxs)
+    })
+
+  def sliding(n: Int, step: Int = 1): Repr[immutable.Seq[Ctx], immutable.Seq[Out]] =
+    via(flow.sliding(n, step).map { elsWithContext ⇒
+      val (els, ctxs) = elsWithContext.unzip
+      (els, ctxs)
+    })
+
+  def mapConcat[Out2](f: Out ⇒ immutable.Iterable[Out2]): Repr[Ctx, Out2] = statefulMapConcat(() ⇒ f)
+
+  def statefulMapConcat[Out2](f: () ⇒ Out ⇒ immutable.Iterable[Out2]): Repr[Ctx, Out2] = {
+    val fCtx: () ⇒ ((Out, Ctx)) ⇒ immutable.Iterable[(Out2, Ctx)] = { () ⇒ elWithContext ⇒
+      val (el, ctx) = elWithContext
+      f()(el).map(o ⇒ (o, ctx))
+    }
+    via(flow.statefulMapConcat(fCtx))
+  }
+
+  def mapContext[Ctx2](f: Ctx ⇒ Ctx2): Repr[Ctx2, Out] =
+    via(flow.map { case (e, ctx) ⇒ (e, f(ctx)) })
+
+  def endContextPropagation: Prov[Ctx, Out]
+
+  private[akka] def flow[T, C]: Flow[(T, C), (T, C), NotUsed] = Flow[(T, C)]
+}
+
+/**
+ * API MAY CHANGE
+ */
+@ApiMayChange
+object FlowWithContext {
+  def apply[Ctx, In]: FlowWithContext[Ctx, In, Ctx, In, akka.NotUsed] = {
+    val under = Flow[(In, Ctx)]
+    new FlowWithContext[Ctx, In, Ctx, In, akka.NotUsed](under, under.traversalBuilder, under.shape)
+  }
+  def from[CI, I, CO, O, M](flow: Flow[(I, CI), (O, CO), M]) = new FlowWithContext(flow, flow.traversalBuilder, flow.shape)
+}
+
+/**
+ * API MAY CHANGE
+ */
+@ApiMayChange
+final class FlowWithContext[-CtxIn, -In, +CtxOut, +Out, +Mat](
+  underlying:                    Flow[(In, CtxIn), (Out, CtxOut), Mat],
+  override val traversalBuilder: LinearTraversalBuilder,
+  override val shape:            FlowShape[(In, CtxIn), (Out, CtxOut)]
+) extends FlowWithContextOps[CtxOut, Out, Mat] with Graph[FlowShape[(In, CtxIn), (Out, CtxOut)], Mat] {
+
+  override def withAttributes(attr: Attributes): Repr[CtxOut, Out] = new FlowWithContext(underlying, traversalBuilder.setAttributes(attr), shape)
+
+  override type Repr[+C, +O] = FlowWithContext[CtxIn @uncheckedVariance, In @uncheckedVariance, C, O, Mat @uncheckedVariance]
+  override type Prov[+C, +O] = Flow[(In @uncheckedVariance, CtxIn @uncheckedVariance), (O, C), Mat @uncheckedVariance]
+
+  override def via[Ctx2, Out2, Mat2](viaFlow: Graph[FlowShape[(Out, CtxOut), (Out2, Ctx2)], Mat2]): Repr[Ctx2, Out2] = from(underlying.via(viaFlow))
+
+  def to[Mat2](sink: Graph[SinkShape[(Out, CtxOut)], Mat2]): Sink[(In, CtxIn), Mat] = underlying.toMat(sink)(Keep.left)
```

Comment:
Are the `to` and `toMat` methods here particularly useful? It seems those are rarely used even on regular flows. They can be easily replaced by `endContextPropagation.to` and `endContextPropagation.toMat` so I would vote for removing them.

Reply:
They are used in our stuff, it's a little shortcut, but it is really in the way, they can be removed.